demos/common/hfsm/az_iot_hfsm.c

```diff
@@ -0,0 +1,324 @@
+      (child_state == provisioning) ||
+#endif
+      (child_state == hub) || (child_state == idle)
+    )
+  {
+    parent_state = azure_iot;
+  }
+  else {
+    // Unknown state.
+    az_iot_hfsm_pal_critical_error(NULL);
+    parent_state = NULL;
+  }
+
+  return parent_state;
+}
+
+// AzureIoT
+static int32_t azure_iot(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+  int32_t operation_msec;
+
+  az_iot_hfsm_type* this_iothfsm = (az_iot_hfsm_type*)me;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+      LogInfo( ("AzureIoT: Entry") );
+      this_iothfsm->_use_secondary_credentials = false;
+      break;
+
+    case AZ_IOT_ERROR:
+      LogInfo( ("AzureIoT: AZ_IOT_ERROR") );
+      operation_msec = az_hfsm_pal_timer_get_milliseconds() - this_iothfsm->_start_time_msec;
+
+      if (this_iothfsm->_retry_attempt < INT16_MAX)
+      {
+        this_iothfsm->_retry_attempt++;
+      }
+
+      az_iot_hfsm_event_data_error* error_data = (az_iot_hfsm_event_data_error*)(event.data);
+
+      bool should_retry = false;
+      if (error_data->type == AZ_IOT_ERROR_TYPE_SERVICE)
+      {
+        should_retry = az_iot_status_retriable(error_data->iot_status);
+      }
+      else if (error_data->type == AZ_IOT_ERROR_TYPE_NETWORK)
+      {
+        should_retry = true;
+      }
+
+      if (should_retry)
+      {
+        int32_t random_jitter_msec = az_iot_hfsm_pal_get_random_jitter_msec(me);
+
+        int32_t retry_delay_msec = az_iot_calculate_retry_delay(
+          operation_msec, 
+          this_iothfsm->_retry_attempt,
+          AZ_IOT_HFSM_MIN_RETRY_DELAY_MSEC,
+          AZ_IOT_HFSM_MAX_RETRY_DELAY_MSEC,
+          random_jitter_msec);
+
+        LogWarn( ("AzureIoT: AZ_IOT_ERROR retrying after %" PRId32 "ms", retry_delay_msec) );
+        ret = az_hfsm_pal_timer_start(me, this_iothfsm->_timer_handle, retry_delay_msec, true);
+      }
+      else
+      {
+        this_iothfsm->_use_secondary_credentials = !this_iothfsm->_use_secondary_credentials;
+        this_iothfsm->_start_time_msec = az_hfsm_pal_timer_get_milliseconds();
+
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+          ret = az_hfsm_post_event(this_iothfsm->_provisioning_hfsm, az_hfsm_event_az_iot_start);
+          ret = az_hfsm_transition_substate(me, azure_iot, provisioning);
+#else
+          ret = az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start);
+          ret = az_hfsm_transition_substate(me, azure_iot, hub);
+#endif
+      }
+      break;
+
+    case AZ_HFSM_EXIT:
+    case AZ_HFSM_ERROR:
+    case AZ_HFSM_TIMEOUT:
+    default:
+      LogInfo( ("AzureIoT: PANIC!") );
+      az_iot_hfsm_pal_critical_error(me);
+      break;
+  }
+
+  return ret;
+}
+
+static int32_t idle(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+  az_iot_hfsm_type* this_iothfsm = (az_iot_hfsm_type*)me;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+      LogInfo( ("idle: Entry") );
```

Comment:
this is going to bubble up in the log file with other traces, so

"idle:: should be "hfsm idle:" ?

Reply:
Standardized az_log:

```C
      if (_az_LOG_SHOULD_WRITE(AZ_LOG_HFSM_ENTER))
      {
        // AZ_LOG_HFSM_ENTER - common events (enter/exit/timeout/error) are standardized across state machines.
        // The string span should be standardized to represent the full path to the state (including any superstates and the state machine as root)
        _az_LOG_WRITE(AZ_LOG_HFSM_ENTER, AZ_SPAN_FROM_STR("az_iot_hfsm/azure_iot/idle"));
      }
      break;
```